### PR TITLE
[Feature][Model] Add Dots3 Note support

### PR DESCRIPTION
**Pull request by QwertyJack**
### What this PR does / why we need it?

Adds Dots3 Note support (text / image / audio / video + MTP3) on Ascend main.

- `vllm_ascend/models/dots3_note/`: Ascend model, vision/audio towers, mixed full/sliding MLA, MTP with its dedicated embedding, and the `Dots3NoteForCausalLM` / `Dots3NoteMTPModel` registrations.
- `vllm_ascend/attention/{mla_v1,sfa_v1}.py`, `vllm_ascend/ops/mla.py`: layer-local RoPE, rope-only K layernorm, q/kv LoRA rescale and headwise/elementwise output gating; paged sliding-window MLA decode (`models/dots3_note/sliding_window_mla.py`) plus an eager prefill fallback for layers where `qk_nope_head_dim != v_head_dim`.
- `vllm_ascend/worker/model_runner_v1.py`, `vllm_ascend/patch/platform/patch_kv_cache_utils.py`: sliding-window MLA cache spec, per-spec `store_on_host` sparse-offload routing, and KV-cache grouping when only one MLA block size exists.
- `vllm_ascend/patch/`: PyAV video/audio decoding on the upstream sampled-frame timeline, raw video/audio passthrough, question + output-budget injection, and Dots3 Note video defaults.
- `vllm_ascend/spec_decode/llm_base_proposer.py`: MTP keeps its checkpoint-provided `embed_tokens` instead of sharing the target model's.

### Does this PR introduce _any_ user-facing change?

Dots3 Note can serve text, images, audio, and videos on Ascend. Native video requires `av`. Validated configuration: model runner v1 (`VLLM_USE_V2_MODEL_RUNNER=0`), `FULL_DECODE_ONLY`, and MTP3 with an eager draft.

### How was this patch tested?

- Unit tests of all touched test files (241 tests), formatting and changed-file mypy pass; CI on `800f5b052` is green (26 success, 4 skipped).
- `/models/dots3-note-prev` BF16, Ascend 910_93, TP16 + EP, runner v1, `FULL_DECODE_ONLY`, MTP3: 10/10 HTTP requests (5 text, 3 image, 2 video) and 2/2 concurrent text requests, no failures or truncations, graph capture and replay confirmed. MTP accepted 2006/2643 drafted tokens (773/676/557 by position).
- Video rendering matches the official Transformers reference on both test videos: 84/75 text/image/audio parts and 7521/7551 rendered token IDs, with BF16 image tensors and FP32 log-mel tensors identical.
- Not covered: 128K / bs16 capacity and performance.

### Co-authors

- @hugler (`huguoliang@huawei.com`)
- @mdai1997 (`184821467@qq.com`)
- @mjlovesz (`lcj200202513@163.com`)

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by QwertyJack** [comment]
/rerun
[Bot]: rerun completed.

**Rerun:**
- E2E

File: vllm_ascend/patch/dots3_note_mtp.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    ) -> Iterator[tuple[str, torch.Tensor]]:
+        first_layer = self.model.mtp_start_layer_idx
+        share_layers = self.model.mtp_head_sharing == "full"
+        dedicated_embeddings = getattr(self.config, "use_dedicated_mtp_embeddings", False)
+
+        for name, weight in weights:
+            if dedicated_embeddings and name == "model.embed_tokens.weight":
+                continue
+            if dedicated_embeddings and name == "model.mtp.embed_tokens.weight":
+                name = f"model.layers.{first_layer}.embed_tokens.weight"
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/patch/dots3_note_mtp.py]: Keeping this mapping as written. `Dots3NoteMTPModel.load_weights()` delegates to `DeepSeekMTP.load_weights()`, whose `_rewrite_spec_layer_name()` rewrites `model.layers.{spec_layer}.embed_tokens` to the model-level `model.embed_tokens` because `embed_tokens` is a shared weight. Mapping directly to `model.embed_tokens` before the base loader would lose the speculative-layer index and the weight would be skipped. `test_dots3_note_mtp_reuses_deepseek_loader_with_dots3_note_weight_layout` covers this checkpoint-to-loader mapping.

File: vllm_ascend/patch/dots3_note_video.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+            best_frame = None
+            best_distance = float("inf")
+            for frame in container.decode(stream):
+                frame_time = float(frame.pts * frame.time_base) if frame.pts else 0.0
+                distance = abs(frame_time - target_time)
+                if distance < best_distance:
+                    best_frame = frame
+                    best_distance = distance
+                if frame_time >= target_time:
+                    break
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/patch/dots3_note_video.py]: Fixed in `a9026ca5b`. The decoder now calls `stream.codec_context.flush_buffers()` after every seek and uses `frame.time` with a safe PTS/time-base fallback. Added a focused regression test that verifies the seek, flush, and `frame.time` path.

File: vllm_ascend/models/dots3_note/audio.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+        N_FFT,
+        HOP_LENGTH,
+        window=_hann_window(),
+        return_complex=True,
+    )
+    magnitudes = stft[..., :-1].abs().square()
+    mel_spec = _mel_filters() @ magnitudes
+    log_spec = torch.clamp(mel_spec, min=1e-10).log10()
+    log_spec = torch.maximum(log_spec, log_spec.max() - 8.0)
+    return (log_spec + 4.0) / 4.0
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/models/dots3_note/audio.py:61]: Keeping the current device placement. This preprocessing path constructs each waveform with `torch.as_tensor(..., dtype=torch.float32)` without a device, so STFT and mel projection are intentionally CPU preprocessing. Only the resulting `audio_features` tensor is moved to the audio tower device/dtype during model execution. Moving cached CPU filters/windows per call is unnecessary for this contract.

File: vllm_ascend/models/dots3_note/audio.py [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+class Dots3NoteAudioRMSNorm(nn.Module):
+    def __init__(self, hidden_size: int, eps: float = 1e-6) -> None:
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(hidden_size))
+        self.eps = eps
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        variance = hidden_states.pow(2).mean(-1, keepdim=True)
+        hidden_states = hidden_states * torch.rsqrt(variance + self.eps)
+        return self.weight * hidden_states
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/models/dots3_note/audio.py]: Keeping the checkpoint/native implementation semantics here. The official Dots3 Note Transformers implementation computes this audio RMSNorm in the input dtype and returns `weight * hidden_states` without an FP32 round trip. The final BF16 TP16/EP16 real-weight text and image smoke tests pass. Changing its numeric contract should be a separate accuracy change backed by matched parity evidence.

File: vllm_ascend/models/dots3_note/multimodal.py [hunk 5]
```diff
@@ -0,0 +1,1863 @@
...
+            )
+            wpos_ids = wpos_ids.permute(0, 2, 1, 3).flatten()
+            pos_ids.append(torch.stack([hpos_ids, wpos_ids], dim=-1).repeat(t, 1))
+        return pos_ids
+
+    def rot_pos_emb(self, grid_thw: list[list[int]]) -> torch.Tensor:
+        pos_ids = torch.cat(self.get_pos_ids_by_grid(grid_thw), dim=0)
+        max_grid_size = max(max(h, w) for _, h, w in grid_thw)
+        rotary_pos_emb_full = self.rotary_pos_emb(max_grid_size)
+        return rotary_pos_emb_full[pos_ids].flatten(1)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/models/dots3_note/multimodal.py:343]: Fixed in `a9026ca5b`. `pos_ids` is now moved to `rotary_pos_emb_full.device` immediately before indexing.

File: vllm_ascend/attention/sfa_v1.py [hunk 6]
```diff
@@ -1704,6 +1771,25 @@ def forward(
...
+            if self.sdpa_gate_type == "elementwise":
+                gated_output = gated_output * gate.view_as(gated_output)
+            else:
+                attn_shape = gated_output.shape
+                gated_output = gated_output.view(
+                    -1,
+                    self.local_num_heads,
+                    self.v_head_dim,
+                ) * gate.view(-1, self.local_num_heads, 1)
+                gated_output = gated_output.view(attn_shape)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/attention/sfa_v1.py]: Fixed in `a9026ca5b`; the SFA headwise gate path now uses `reshape` and restores the original shape with `reshape`.

File: vllm_ascend/patch/worker/patch_dots3_note.py [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+        q = self.q_b_proj(q_c)[0]
+    else:
+        kv_lora = self.kv_a_proj_with_mqa(hidden_states)[0]
+        q = self.q_proj(hidden_states)[0]
+
+    kv_c, k_pe = kv_lora.split([self.kv_lora_rank, self.qk_rope_head_dim], dim=-1)
+    kv_c_normed = self.kv_a_layernorm(kv_c)
+    if self.apply_mla_qkv_lora_rescale:
+        kv_c_normed = kv_c_normed * math.sqrt(self.hidden_size / self.kv_lora_rank)
+    q = q.view(-1, self.num_heads, self.qk_head_dim)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/patch/worker/patch_dots3_note.py]: Fixed in `a9026ca5b`; the potentially non-contiguous query tensor now uses `reshape`.

File: vllm_ascend/patch/worker/patch_dots3_note.py [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+        if self.sdpa_gate_type == "headwise" and gate.shape[-1] != self.num_heads:
+            rank = get_tensor_model_parallel_rank()
+            gate = gate.narrow(-1, rank * self.num_heads, self.num_heads)
+        gate = torch.sigmoid(gate.float()).to(attn_out.dtype)
+        if self.sdpa_gate_type == "elementwise":
+            attn_out = attn_out * gate.view_as(attn_out)
+        else:
+            attn_shape = attn_out.shape
+            attn_out = attn_out.view(-1, self.num_heads, self.v_head_dim) * gate.view(-1, self.num_heads, 1)
+            attn_out = attn_out.view(attn_shape)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/patch/worker/patch_dots3_note.py]: Fixed in `a9026ca5b`; the headwise gate path now uses `reshape` for both attention output and gate tensors.

File: vllm_ascend/models/dots3_note/audio.py [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+                        self.rotary_dim,
+                        2,
+                        dtype=torch.float32,
+                        device=device,
+                    )
+                    / max(self.rotary_dim, 1)
+                )
+            )
+        else:
+            inv_freq = self.inv_freq.to(device=device)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/models/dots3_note/audio.py:160]: Keeping the reconstruction. The non-FP32 branch exists because module-wide dtype conversion can already have downcast the `inv_freq` buffer; casting those rounded values back to FP32 cannot recover the lost frequency precision. Recomputing from `rope_theta` and `rotary_dim` restores the intended FP32 values before the autocast-disabled matmul.

File: vllm_ascend/attention/mla_v1.py [hunk 10]
```diff
@@ -1763,9 +2207,29 @@ def forward(
...
+            if self.sdpa_gate_type == "headwise" and gate.shape[-1] != self.num_heads:
+                rank = get_tensor_model_parallel_rank()
+                gate = gate.narrow(-1, rank * self.num_heads, self.num_heads)
+            gate = torch.sigmoid(gate.float()).to(attn_output.dtype)
+            if self.sdpa_gate_type == "elementwise":
+                attn_output = attn_output * gate.view_as(attn_output)
+            else:
+                attn_shape = attn_output.shape
+                attn_output = attn_output.view(-1, self.num_heads, self.v_head_dim) * gate.view(-1, self.num_heads, 1)
+                attn_output = attn_output.view(attn_shape)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/attention/mla_v1.py]: Fixed in `a9026ca5b`; the MLA headwise gate path now uses `reshape` and restores the original shape with `reshape`.

File: vllm_ascend/patch/worker/patch_dots3_note.py [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+    **_: object,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    del hidden_states
+    logit_scores = gating_output.sigmoid()
+    scores_for_choice = logit_scores + e_score_correction_bias.unsqueeze(0)
+    _, topk_idx = torch.topk(scores_for_choice, k=topk, dim=-1, sorted=False)
+    topk_weight = logit_scores.gather(1, topk_idx)
+    if renormalize:
+        topk_weight = topk_weight / topk_weight.sum(dim=-1, keepdim=True)
+    routed_scaling = 1 / topk_weight.norm(dim=-1, keepdim=True) if use_dynamic_rsf else routed_scaling_factor
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/patch/worker/patch_dots3_note.py]: Fixed in `a9026ca5b` with `1e-20`, matching the official Dots3 Note Transformers dynamic-RSF denominator. Added a regression test with zero routing weights that verifies the output remains finite.

File: vllm_ascend/models/dots3_note/multimodal.py [hunk 12]
```diff
@@ -0,0 +1,1863 @@
...
+            selected_mask = topk_indices == expert_idx
+            if not selected_mask.any():
+                continue
+            n_idx, top = torch.where(selected_mask)
+            expert_output = expert(x_flat[n_idx].contiguous())
+            contrib = expert_output * routed_weights[n_idx, top].unsqueeze(-1)
+            aggregated_output[n_idx] = aggregated_output[n_idx] + contrib
+            aggregated_gate[n_idx] = aggregated_gate[n_idx] + routed_weights[n_idx, top]
+
+        return aggregated_output / (aggregated_gate.unsqueeze(-1) + epsilon)
```
[the remark this answers is not in the document]
Comment (QwertyJack) [reply vllm_ascend/models/dots3_note/multimodal.py:217]: Keeping the 2D return contract. The vision patch embed flattens inputs to `[-1, embed_dim]`, and all subsequent Dots3 Note vision blocks, including this MoE, operate on that 2D token matrix. The official implementation likewise flattens before the vision MoE and returns the 2D result. Reshaping to an arbitrary caller shape would only mask an unsupported input.
